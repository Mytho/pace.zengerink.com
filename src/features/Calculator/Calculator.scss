@import '../../App.scss';

@mixin calculator-row {
    display: flex;
    justify-content: space-between;
    padding: 13px 0 13px 13px;
    border-bottom: 1px solid $light-grey;
    background-color: $white;

    &__label {
        color: $medium-grey;
    }

    &__input {
        flex-basis: 180px;
    }

    &__separator,
    &__value-label {
        float: left;
        color: $dark-grey;
    }

    &__separator {
        margin: 0 1px;
    }

    &__value-label {
        margin: 0 7px 0 4px;
    }

    &:last-child {
        border-bottom: 0;
        border-bottom-left-radius: $border-radius;
        border-bottom-right-radius: $border-radius;
    }
}

.calculator {
    width: 280px;
    margin: 40px auto;
    border-radius: $border-radius;
    box-shadow: $box-shadow;

    &__header {
        display: flex;
        justify-content: space-between;
        padding: 13px;
        background-color: $light-grey;
        color: $medium-light-grey;
        border-top-left-radius: $border-radius;
        border-top-right-radius: $border-radius;
    }

    &__header div {
        margin-top: 2px;
    }

    &__header button {
        padding: 4px;
        color: $medium-light-grey;
        border: 0;
        border-radius: 3px;
        background-color: $lighter-grey;
        cursor: pointer;

        &:focus {
            outline: none;
        }

        &:active,
        &:hover {
            color: $dark-highlight;
            background-color: $light-highlight;
        }
    }
}
