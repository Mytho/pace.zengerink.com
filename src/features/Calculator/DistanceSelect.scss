@import './Calculator.scss';

.distance-select {
    @include calculator-row;
}