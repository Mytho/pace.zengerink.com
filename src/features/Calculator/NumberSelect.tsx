import React, { useEffect, useState } from 'react'
import classNames from 'classnames'
import './NumberSelect.scss'

interface NumberSelectProps {
  name: string
  value: number
  onChange: (value: number) => void
  max: string
  pad?: string
}

const NumberSelect = (props: NumberSelectProps) => {
  const [isActive, setIsActive] = useState(false)

  const displayValue = (value: number): string => {
    return props.pad ? value.toString().padStart(parseInt(props.pad, 10), '0') : value.toString()
  }

  const handleOptionClick = (event: React.MouseEvent<HTMLElement>) => {
    const value = event.currentTarget.getAttribute('data-value')
    if (value === null) return
    setIsActive(false)
    props.onChange(parseInt(value, 10))
  }

  const handleMouseDownOutside = (event: globalThis.MouseEvent | globalThis.TouchEvent) => {
    const target = event.target as HTMLElement
    if (!target.classList.contains('number-select__option') && !target.classList.contains('number-select__value')) {
      setIsActive(false)
    }
  }

  const createRenderedOptions = () => {
    const options = []

    for (let i = 0; i <= parseInt(props.max, 10); i++) {
      options.push({ displayValue: displayValue(i), value: i })
    }

    return options.map((option) => {
      return (
        <div
          className={classNames({
            'number-select__option': true,
            'number-select__option--selected': option.value === props.value,
          })}
          key={option.displayValue}
          data-value={option.value}
          onClick={handleOptionClick}
        >
          {option.displayValue}
        </div>
      )
    })
  }

  useEffect(() => {
    document.addEventListener('mousedown', handleMouseDownOutside, false)
    return () => {
      document.removeEventListener('mousedown', handleMouseDownOutside, false)
    }
  })

  return (
    <div className={classNames({ 'number-select': true, 'number-select--active': isActive })}>
      <div className="number-select__wrapper">
        <div className="number-select__value" onClick={() => setIsActive((wasActive) => !wasActive)}>
          {displayValue(props.value)}
        </div>
        <div className="number-select__options">{createRenderedOptions()}</div>
      </div>
    </div>
  )
}

export default NumberSelect
