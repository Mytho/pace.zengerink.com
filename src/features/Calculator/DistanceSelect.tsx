import React, { ReactElement } from 'react'
import NumberSelect from './NumberSelect'
import './DistanceSelect.scss'

type DistanceFragmentName = 'km' | 'm'

interface DistanceFragments {
  km: number
  m: number
}

interface DistanceSelectProps {
  label: string
  value: number
  onChange: (value: number) => void
}

const DistanceSelect = (props: DistanceSelectProps): ReactElement => {
  const toMeters = (fragments: DistanceFragments): number => fragments.km * 1000 + fragments.m

  const toDistanceFragments = (meters: number): DistanceFragments => ({
    km: Math.floor(meters / 1000),
    m: meters % 1000,
  })

  const handleChange = (fragmentName: DistanceFragmentName, value: number) => {
    const fragments = toDistanceFragments(props.value)
    fragments[fragmentName] = value
    props.onChange(toMeters(fragments))
  }

  const distanceFragments = toDistanceFragments(props.value)

  return (
    <div className="distance-select">
      <div className="distance-select__label">{props.label}</div>
      <div className="distance-select__input">
        <NumberSelect
          name="km"
          value={distanceFragments.km}
          max="100"
          onChange={(value) => handleChange('km', value)}
        />
        <span className="distance-select__separator">.</span>
        <NumberSelect
          name="m"
          value={distanceFragments.m}
          max="999"
          pad="3"
          onChange={(value) => handleChange('m', value)}
        />
        <span className="distance-select__value-label">m</span>
      </div>
    </div>
  )
}

export default DistanceSelect
