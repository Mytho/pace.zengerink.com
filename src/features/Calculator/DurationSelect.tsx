import React from 'react'
import NumberSelect from './NumberSelect'
import './DurationSelect.scss'

export enum DurationFragment {
  Hours = 'hours',
  Minutes = 'minutes',
  Seconds = 'seconds',
}

interface DurationSelectProps {
  label: string
  value: number
  onChange: (value: number) => void
}

interface Duration {
  [key: string]: number
}

export default class DurationSelect extends React.Component<DurationSelectProps, {}> {
  handleChange(prop: string, value: number) {
    const duration = this.toDuration(this.props.value)
    duration[prop] = value
    this.props.onChange(this.toSeconds(duration))
  }

  toSeconds(duration: Duration): number {
    const hours = duration[DurationFragment.Hours] * 3600
    const minutes = duration[DurationFragment.Minutes] * 60

    return hours + minutes + duration[DurationFragment.Seconds]
  }

  toDuration(seconds: number): Duration {
    const date = new Date(seconds * 1000)

    return {
      [DurationFragment.Hours]: date.getUTCHours(),
      [DurationFragment.Minutes]: date.getUTCMinutes(),
      [DurationFragment.Seconds]: date.getSeconds(),
    }
  }

  render() {
    const duration = this.toDuration(this.props.value)

    return (
      <div className="duration-select">
        <div className="duration-select__label">{this.props.label}</div>
        <div className="duration-select__input">
          <NumberSelect
            name={DurationFragment.Hours}
            value={duration[DurationFragment.Hours]}
            max="23"
            onChange={this.handleChange.bind(this, DurationFragment.Hours)}
          />
          <span className="duration-select__value-label">hrs</span>
          <NumberSelect
            name={DurationFragment.Minutes}
            value={duration[DurationFragment.Minutes]}
            max="59"
            pad="2"
            onChange={this.handleChange.bind(this, DurationFragment.Minutes)}
          />
          <span className="duration-select__value-label">min</span>
          <NumberSelect
            name={DurationFragment.Seconds}
            value={duration[DurationFragment.Seconds]}
            max="59"
            pad="2"
            onChange={this.handleChange.bind(this, DurationFragment.Seconds)}
          />
          <span className="duration-select__value-label">sec</span>
        </div>
      </div>
    )
  }
}
