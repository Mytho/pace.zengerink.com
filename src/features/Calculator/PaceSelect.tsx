import React, { ReactElement } from 'react'
import NumberSelect from './NumberSelect'
import './PaceSelect.scss'

type PaceFragmentName = 'minutes' | 'seconds'

interface PaceSelectProps {
  label: string
  value: number
  onChange: (value: number) => void
}

interface PaceFragments {
  minutes: number
  seconds: number
}

const toSeconds = (pace: PaceFragments): number => pace.minutes * 60 + pace.seconds
const toPace = (seconds: number): PaceFragments => ({ minutes: Math.floor(seconds / 60), seconds: seconds % 60 })

const PaceSelect = (props: PaceSelectProps): ReactElement => {
  const handleChange = (fragmentName: PaceFragmentName, value: number) => {
    const fragments = toPace(props.value)
    fragments[fragmentName] = value
    props.onChange(toSeconds(fragments))
  }

  const paceFragments = toPace(props.value)

  return (
    <div className="pace-select">
      <div className="pace-select__label">{props.label}</div>
      <div className="pace-select__input">
        <NumberSelect
          name="minutes"
          value={paceFragments.minutes}
          max="59"
          onChange={(value) => handleChange('minutes', value)}
        />
        <span className="pace-select__separator">:</span>
        <NumberSelect
          name="seconds"
          value={paceFragments.seconds}
          max="59"
          pad="2"
          onChange={(value) => handleChange('seconds', value)}
        />
        <span className="pace-select__value-label">min/km</span>
      </div>
    </div>
  )
}

export default PaceSelect
