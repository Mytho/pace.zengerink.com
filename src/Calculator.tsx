import React from 'react'
import DurationSelect from './DurationSelect'
import DistanceSelect from './DistanceSelect'

enum CalculatorValue {
    Seconds = 'seconds',
    Meters = 'meters'
}

interface CalculatorProps {}

interface CalculatorState {
    [key: string]: number
}

export default class Calculator extends React.Component<CalculatorProps, CalculatorState> {
    constructor(props: any) {
        super(props)

        this.state = {
            [CalculatorValue.Seconds]: 0,
            [CalculatorValue.Meters]: 0
        }
    }

    handleChange(prop: CalculatorValue, value: number) {
        this.setState({
            [prop]: value,
        })
    }

    render() {
        return (
            <div className="calculator">
                <div className="calculator__header">Pace Calculator</div>
                <DurationSelect value={this.state[CalculatorValue.Seconds]} onChange={this.handleChange.bind(this, CalculatorValue.Seconds)} />
                <DistanceSelect value={this.state[CalculatorValue.Meters]} onChange={this.handleChange.bind(this, CalculatorValue.Meters)} />
            </div>
        )
    }
}