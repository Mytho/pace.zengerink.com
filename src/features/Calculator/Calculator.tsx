import React, { ReactElement, useReducer } from 'react'
import DurationSelect from './DurationSelect'
import DistanceSelect from './DistanceSelect'
import PaceSelect from './PaceSelect'
import './Calculator.scss'

type CalculatorField = 'duration' | 'distance' | 'pace'

interface CalculatorState {
  duration: number
  distance: number
  pace: number
  changedFields: CalculatorField[]
}

interface CalculatorChangePayload {
  field: CalculatorField
  value: number
}

interface CalculatorAction {
  type: string
  payload?: CalculatorChangePayload
}

const calculateDistance = (duration: number, pace: number) => {
  const durationInMinutes = duration / 60
  const paceInMinutes = pace / 60
  const paceInKilometers = durationInMinutes / paceInMinutes
  return Math.round(paceInKilometers * 1000)
}

const calculateDuration = (distance: number, pace: number) => {
  const distanceInKilometers = distance / 1000
  const paceInMinutes = pace / 60
  const durationInMinutes = distanceInKilometers * paceInMinutes
  return Math.round(durationInMinutes * 60)
}

const calculatePace = (duration: number, distance: number) => {
  const distanceInKilometers = distance / 1000
  const durationInMinutes = duration / 60
  const paceInMinutes = durationInMinutes / distanceInKilometers
  return Math.round(paceInMinutes * 60)
}

const Calculator = (): ReactElement => {
  const defaultState = {
    duration: 0,
    distance: 0,
    pace: 0,
    changedFields: [],
  } as CalculatorState

  const reducer = (prevState: CalculatorState, action: CalculatorAction): CalculatorState => {
    switch (action.type) {
      case 'calculator/clear':
        return defaultState

      case 'calculator/change':
        const field = action.payload?.field ? action.payload?.field : ''
        const value = action.payload?.value

        const getValue = (fieldName: CalculatorField) => {
          const val = field === fieldName ? value : prevState[fieldName]
          return val ? val : 0
        }

        const duration = getValue('duration')
        const distance = getValue('distance')
        const pace = getValue('pace')

        const changedFields = [
          ...prevState.changedFields.filter((f: CalculatorField) => f !== field).slice(-1),
          field as CalculatorField,
        ]

        const shouldCalculate = (fieldName: CalculatorField): boolean => {
          if (changedFields.length < 2) return false
          return !changedFields.find((changedField: CalculatorField) => changedField === fieldName)
        }

        return {
          duration: shouldCalculate('duration') ? calculateDuration(distance, pace) : duration,
          distance: shouldCalculate('distance') ? calculateDistance(duration, pace) : distance,
          pace: shouldCalculate('pace') ? calculatePace(duration, distance) : pace,
          changedFields,
        }

      default:
        throw new Error()
    }
  }

  const [state, dispatch] = useReducer(reducer, defaultState)

  return (
    <div className="calculator">
      <div className="calculator__header">
        <div>Pace Calculator</div>
        <button onClick={() => dispatch({ type: 'calculator/clear' })}>Clear</button>
      </div>
      <DurationSelect
        label="Time"
        value={state?.duration}
        onChange={(value) => dispatch({ type: 'calculator/change', payload: { field: 'duration', value } })}
      />
      <DistanceSelect
        label="Distance"
        value={state?.distance}
        onChange={(value) => dispatch({ type: 'calculator/change', payload: { field: 'distance', value } })}
      />
      <PaceSelect
        label="Pace"
        value={state?.pace}
        onChange={(value) => dispatch({ type: 'calculator/change', payload: { field: 'pace', value } })}
      />
    </div>
  )
}

export default Calculator
