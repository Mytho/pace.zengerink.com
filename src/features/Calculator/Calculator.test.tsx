import React from 'react'
import { mount, ReactWrapper, shallow } from 'enzyme'
import { shallowToJson } from 'enzyme-to-json'
import { act } from 'react-dom/test-utils'
import Calculator from './Calculator'

const duration = 600 // 10 min
const distance = 2000 // 2 km
const pace = 300 // 5 min/km

const change = (wrapper: ReactWrapper, label: string, value: number): ReactWrapper => {
  act(() => {
    wrapper.find(`[label="${label}"]`).props().onChange!(value as any)
  })
  return wrapper.update()
}

const validate = (wrapper: ReactWrapper, label: string, value: number) => {
  expect(wrapper.find(`[label="${label}"]`).props().value).toBe(value)
}

describe('render', () => {
  it('renders correctly', () => {
    const json = shallowToJson(shallow(<Calculator />))
    expect(json).toMatchSnapshot()
  })
})

describe('calculate', () => {
  it('calculates pace when duration and distance have been set', () => {
    let wrapper = mount(<Calculator />)

    wrapper = change(wrapper, 'Time', duration)
    wrapper = change(wrapper, 'Distance', distance)

    validate(wrapper, 'Time', duration)
    validate(wrapper, 'Distance', distance)
    validate(wrapper, 'Pace', pace)
  })

  it('calculates duration when distance and pace have been set', () => {
    let wrapper = mount(<Calculator />)

    wrapper = change(wrapper, 'Distance', distance)
    wrapper = change(wrapper, 'Pace', pace)

    validate(wrapper, 'Time', duration)
    validate(wrapper, 'Distance', distance)
    validate(wrapper, 'Pace', pace)
  })

  it('calculates distance when duration and pace have been set', () => {
    let wrapper = mount(<Calculator />)

    wrapper = change(wrapper, 'Time', duration)
    wrapper = change(wrapper, 'Pace', pace)

    validate(wrapper, 'Time', duration)
    validate(wrapper, 'Distance', distance)
    validate(wrapper, 'Pace', pace)
  })

  it('calculates the value of the last unchanged select', () => {
    let wrapper = mount(<Calculator />)

    wrapper = change(wrapper, 'Time', duration)
    wrapper = change(wrapper, 'Distance', distance)
    wrapper = change(wrapper, 'Pace', pace)

    validate(wrapper, 'Time', duration)
    validate(wrapper, 'Distance', distance)
    validate(wrapper, 'Pace', pace)

    const anotherDuration = 300 // 5 min
    const anotherDistance = 5000 // 5 km
    const anotherPace = 60 // 1 min/km

    wrapper = change(wrapper, 'Time', anotherDuration)
    wrapper = change(wrapper, 'Pace', anotherPace)

    validate(wrapper, 'Time', anotherDuration)
    validate(wrapper, 'Distance', anotherDistance)
    validate(wrapper, 'Pace', anotherPace)
  })

  it('does not trigger any calculation when same select is changed twice', () => {
    let wrapper = mount(<Calculator />)

    wrapper = change(wrapper, 'Time', 3600) // 1 hour
    wrapper = change(wrapper, 'Time', 3660) // 1 hour and 1 minute

    validate(wrapper, 'Time', 3660)
    validate(wrapper, 'Distance', 0)
    validate(wrapper, 'Pace', 0)
  })
})

describe('clear', () => {
  it('clears the values when clicked', () => {
    let wrapper = mount(<Calculator />)

    validate(wrapper, 'Time', 0)
    validate(wrapper, 'Distance', 0)
    validate(wrapper, 'Pace', 0)

    wrapper = change(wrapper, 'Time', duration)
    wrapper = change(wrapper, 'Distance', distance)

    validate(wrapper, 'Time', duration)
    validate(wrapper, 'Distance', distance)

    act(() => {
      wrapper.find('button').simulate('click')
    })
    wrapper.update()

    validate(wrapper, 'Time', 0)
    validate(wrapper, 'Distance', 0)
    validate(wrapper, 'Pace', 0)
  })
})
