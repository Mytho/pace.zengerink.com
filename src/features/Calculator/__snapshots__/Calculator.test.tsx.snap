// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render renders correctly 1`] = `
<div
  className="calculator"
>
  <div
    className="calculator__header"
  >
    <div>
      Pace Calculator
    </div>
    <button
      onClick={[Function]}
    >
      Clear
    </button>
  </div>
  <DurationSelect
    label="Time"
    onChange={[Function]}
    value={0}
  />
  <DurationSelect
    label="Distance"
    onChange={[Function]}
    value={0}
  />
  <DurationSelect
    label="Pace"
    onChange={[Function]}
    value={0}
  />
</div>
`;
